### ISSUE TITLE
lib-storage should compute MD5 for parts

### ISSUE THREAD

**Issue by paya-cz** [aws/aws-sdk-js-v3#2673]
### Is your feature request related to a problem? Please describe.
[UploadPart API](https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html) supports specifying MD5/SHA256 to verify each part is received intact. `lib-storage` does not allow me to make use of this feature.

### Describe the solution you'd like
Add a flag to the [Upload constructor](https://github.com/aws/aws-sdk-js-v3/blob/c458481ce07b14017a671603b79111a9958d7f0c/lib/lib-storage/src/Upload.ts#L59) that either enables or disables automatic MD5 computation for each part.

### Describe alternatives you've considered
Fork `lib-storage`.

**Comment by phillycheeze** [comment]
+1 for this. To set an ObjectLock on an uploaded object, AWS requires that the uploaded file or uploaded parts include the `Content-MD5` header. Otherwise, the client-s3 fails with a `MalformedXML` error at the end of the upload.

It looks like this upgrading doc mentions the `computeChecksums` client constructor option was something that didn't transfer from v2 -> v3 and is still marked as "planned".

For anyone reading this, the best workaround is to use v2 of the SDK for now. Do not try to set the Content-MD5 header on the Upload params in this library, since the http requests show it's just copying that hash value and uploading it with every single part which breaks all of the parts.

**Comment by phillycheeze** [comment]
Okay I figured out a workaround using the existing packages in this library. It seems like there is a package called `@aws-sdk/middleware-apply-body-checksum` that most of the commands in the client-s3 use to automatically generate a md5 hash and add the _content-md5_ header. I was able to get this to work via the following:

```
const { S3Client } = require('@aws-sdk/client-s3')
const { getApplyMd5BodyChecksumPlugin } = require('@aws-sdk/middleware-apply-body-checksum')

const client = new S3Client({
  region: 'us-east-2',
})
client.middlewareStack.use(
  getApplyMd5BodyChecksumPlugin(awsClient.config)
)
```

This will inject the existing package into the middleware stack for all S3 client requests and then you can continue to use the Upload class from lib-storage as normal. There could be other types of requests that you don't want this middle to run for, so you can conditionally do the `middlewareStack.use` above before your request and then use `middlewareStack.remove('applyMd5BodyChecksumMiddleware')` to remove it.

After running the workaround above, multipart uploads work like a charm when setting `ObjectLockMode` and `ObjectLockRetainUntilDate` in the upload params.

side note: it seems the Upload class expects a type of _PutObjectCommandInput_ for the params config, but when you give it a stream it proceeds to never actually call "PutObject" since it needs to do a multi-part upload. When you compare AWS's api requirements between _PutObject_ and _CreateMultiPartUpload_ there are some minor differences. One difference is the `Content-MD5` which is expected in the _PutObject_ command but isn't used in the _CreateMultiPartUpload_ command. Looking through the source it seems like some hacks have been made in the lib-storage library to set the _Body_ key to `undefined` as a way around it. It's a little odd to have params be validated against a command input type that it never even calls and it's probably why it took so long to debug this issue since originally I kept getting back _MalformedXML_ 400 status code errors from aws.

**Comment by paya-cz** [comment]
@phillycheeze Excellent find! Hopefully AWS decides to add this as a simple boolean switch to the `Upload` class one day. And updates it to use SHA-2 as well! The fact that this middleware affects every single request is a little worrisome though.

**Comment by bret540** [comment]
@phillycheeze solution didn't work in my case (same malformed xml error).  I'm not using ObjectLockMode options so not sure if this solution requires this feature on a bucket.  (Same code works for smaller files that won't do multipart)
```
const s3Client = new S3Client()
s3Client.middlewareStack.use(
  getApplyMd5BodyChecksumPlugin(s3Client.config)
)
const upload = new Upload({
    params: {
        Bucket: bucketName,
        Key: fileName,
        Body: readableStream
    },
    client: s3Client
})
upload.on('httpUploadProgress', (progress) => {
    console.log(`Progress: ${progress}`)
})
await upload.done()
```

**Comment by phillycheeze** [comment]
@ch3apsh0t are there any other parameters you are giving to the Upload() call? When initially trying to debug this, I originally tried really hard to identify the source of the bad XML itself. I never found out how to do it and the AWS documentation is pretty vague on troubleshooting it.

If it works with smaller files (< the default 5MB part size), that's because the library is sending an api request to AWS with everything together instead of a multi-part upload. If the error is happening at the very end of the uploading (progress logs in the console are printing up until the end), then it's likely AWS stitching everything together and then validating it. The final "part" fails because AWS says... "hold up there, actually something is wrong with this file when we packaged it all up." That's why in this issue thread, the problem is that AWS can't know if the file fails validation until it receives everything and checks the MD5 checksum (which is a requirement for parameters like ObjectLockHold and others).

It's possible your issue is something else instead of the MD5 checksums. If you're certain it's only on multi-part uploads at the end of uploading and not small files, my main recommendation is to look up AWS documentation for parameters required for PutObject calls and the MultiPart calls. If the `params` in your code snippet are truly all you are using, I'm inclined to think the issue is different than the one here and might be coming from your `ReadableStream` whatever that is.

**Comment by phillycheeze** [comment]
I spent several days to figure all of that out and at the end of the day, a MalformedXML error is really a bug with the sdk itself - not user error. I mention this because using v2 of the aws-sdk-js package might save you a bunch of your time and sanity! 😅

**Comment by JacopoBonta** [comment]
> I spent several days to figure all of that out and at the end of the day, a MalformedXML error is really a bug with the sdk itself - not user error. I mention this because using v2 of the aws-sdk-js package might save you a bunch of your time and sanity! 😅

Hi, I probably have a similar issue.
In my case, I'm not even able to log the progress, and the complete multipart upload request terminates with a 400 XML malformed error.

I noticed that the complete multipart request is sent without any `<Part>` inside the XML, which causes it to be malformed, I guess. From [here](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html#API_CompleteMultipartUpload_RequestBody) **Part** is not required but causes the server to return a 400.

IMO the bug lies inside the `lib-storage` as the request is sent without those parts [here](https://github.com/aws/aws-sdk-js-v3/blob/f23158a36f322300014d4ad74b41f9360580409e/lib/lib-storage/src/Upload.ts#L321).
```
      const uploadCompleteParams = {
        ...this.params,
        Body: undefined,
        UploadId: this.uploadId,
        MultipartUpload: {
          Parts: this.uploadedParts, // here `this.uploadedParts` is always an empty array during my tests
        },
      };
      result = await this.client.send(new CompleteMultipartUploadCommand(uploadCompleteParams));
```

**Comment by skypesky** [comment]
related: https://github.com/aws/aws-sdk-js-v3/issues/4321

**Comment by otaviojacobi** [comment]
@JacopoBonta and anyone else still fighting this, I figured it out that several times this issue appears but it is actually being caused by an underlying error on PartUpload (e.g. missing KMS decrypt key). Because of: https://github.com/aws/aws-sdk-js-v3/blob/main/lib/lib-storage/src/Upload.ts#L300-L301 the error gets burried so I would recommend setting `leavePartsOnError` o true and you might find the actual error.

**Comment by xincan1949** [comment]
> @JacopoBonta and anyone else still fighting this, I figured it out that several times this issue appears but it is actually being caused by an underlying error on PartUpload (e.g. missing KMS decrypt key). Because of: https://github.com/aws/aws-sdk-js-v3/blob/main/lib/lib-storage/src/Upload.ts#L300-L301 the error gets burried so I would recommend setting `leavePartsOnError` o true and you might find the actual error when debugging

Add `ETag` ExposeHeader works for me, see this
https://github.com/aws/aws-sdk-js-v3/issues/5057#issuecomment-1671466081

**Comment by piurafunk** [comment]
I was able to get this to work by adding the `ChecksumAlgorithm` property to my `putObject` input:

```javascript
import { ChecksumAlgorithm, S3 } from '@aws-sdk/client-s3'

await s3.putObject({
    Bucket: bucket,
    Key: key,
    Body: body,
    ChecksumAlgorithm: ChecksumAlgorithm.SHA256,
})
```

I found that in [PutObjectCommand.ts#L412](https://github.com/aws/aws-sdk-js-v3/blob/main/clients/client-s3/src/commands/PutObjectCommand.ts#L412) is where it builds the checksum middleware. Following that down, I ended up at [getChecksumAlgorithmforRequest.ts#L21](https://github.com/aws/aws-sdk-js-v3/blob/main/packages/middleware-flexible-checksums/src/getChecksumAlgorithmForRequest.ts#L21). It has this little snipped:

```javascript
if (!requestAlgorithmMember || !input[requestAlgorithmMember]) {
  // Select an algorithm only if request checksum is required.
  return requestChecksumRequired ? defaultAlgorithm : undefined;
}
```

Where `requestAlgorithmMeber` is originally defined in `PutObjectCommand.ts` with the value `'ChecksumAlgorithm'`. Because I (originally) didn't have the `ChecksumAlgorithm` defined in my `putObject()` input, it checks if a checksum is required, and if it isn't it returns `undefined`, thus no checksum is performed at all. Unfortunately, `requestChecksumRequired` is hard-coded to `false` in `PutObjectCommand.ts`, so by default it always returns `undefined`.

So if you put that property in your `putObject()` input, it doesn't do that ternary check and instead continues with this code block:

```javascript
const checksumAlgorithm = input[requestAlgorithmMember];
// Validate that at least one algorithm from customer preference is supported by the SDK.
if (!CLIENT_SUPPORTED_ALGORITHMS.includes(checksumAlgorithm)) {
  throw new Error(
    `The checksum algorithm "${checksumAlgorithm}" is not supported by the client.` +
      ` Select one of ${CLIENT_SUPPORTED_ALGORITHMS}.`
  );
}

return checksumAlgorithm as ChecksumAlgorithm;
```

Which just ensures the algorithm specified is supported, and then casts it as an enum.

---

So basically, define `ChecksumAlgorithm` for your `putObject()` calls, and it'll compute it for you.
